File: 卡码/代码随想录/1/11-图论/Go/并查集/LeetCode_0684.go
// 684. 冗余连接
package main

func findRedundantConnection(edges [][]int) []int {
	n := len(edges)
	father := make([]int, n+1)
	for i := 0; i < n+1; i++ {
		father[i] = i
	}
	var find func(u int) int
	var isSame func(u, v int) bool
	var join func(u, v int)
	find = func(u int) int {
		if u == father[u] {
			return u
		}
		father[u] = find(father[u])
		return father[u]
	}
	isSame = func(u, v int) bool {
		u, v = find(u), find(v)
		return u == v
	}
	join = func(u, v int) {
		u, v = find(u), find(v)
		if u == v {
			return
		}
		father[v] = u
	}
	res := make([]int, 2)
	for _, edge := range edges {
		if isSame(edge[0], edge[1]) {
			copy(res, edge)
		} else {
			join(edge[0], edge[1])
		}
	}
	return res
}

File: 卡码/代码随想录/1/11-图论/Python/并查集/LeetCode_0684.py
#684. 冗余连接
from typing import *

class Solution:
  def __init__(self):
    self.father:List[int]=[]
  def init(self,n:int):
    self.father=[i for i in range(n+1)]
  def find(self,u:int)->int:
    if u==self.father[u]:
      return u
    self.father[u]=self.find(self.father[u])
    return self.father[u]
  def isSame(self,u:int,v:int)->bool:
    u=self.find(u)
    v=self.find(v)
    return u==v
  def join(self,u:int,v:int):
    u=self.find(u)
    v=self.find(v)
    if u==v:
      return
    self.father[v]=u
  def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
    n:int=len(edges)
    self.init(n)
    res:List[int]=[]
    for edge in edges:
      if self.isSame(edge[0],edge[1]):
        res=edge
      else:
        self.join(edge[0],edge[1])
    return res